src/JuliusSweetland.OptiKey/Models/XmlKeyboard.cs

```diff
@@ -0,0 +1,246 @@
+            set {
+                try
+                {
+                    ThicknessConverter thicknessConverter = new ThicknessConverter();
+                    BorderThickness = (Thickness)thicknessConverter.ConvertFromString(value);
+                }
+                catch (System.FormatException)
+                {
+                    Log.ErrorFormat("Cannot interpret \"{0}\" as thickness", value);                
+                }
+            }
+        }        
+        
+        [XmlIgnore]
+        public bool Hidden
+        { get; set; }
+
+        [XmlElement("HideFromKeyboardMenu")]
+        public string HiddenBoolAsString
+        {
+            get { return this.Hidden ? "True" : "False"; }
+            set { this.Hidden = XmlUtils.ConvertToBoolean(value); }
+        }        
+        
+        public static XmlKeyboard ReadFromFile(string inputFilename)
+        {
+            XmlKeyboard keyboard;
+
+            // If no extension given, try ".xml"
+            string ext = Path.GetExtension(inputFilename);
+            bool exists = File.Exists(inputFilename);
+            if (!File.Exists(inputFilename) &&
+                String.IsNullOrEmpty(Path.GetExtension(inputFilename)))
+            {
+                inputFilename += ".xml";
+            } 
+
+            // Read in XML file (may throw)
+            XmlSerializer serializer = new XmlSerializer(typeof(XmlKeyboard));
+            using (FileStream readStream = new FileStream(@inputFilename, FileMode.Open))
+            {
+                keyboard = (XmlKeyboard)serializer.Deserialize(readStream);
+            }
+
+            return keyboard;
+        }
+    }
+
+    public class XmlGrid
+    {
+        public int Rows
+        { get; set; }
+
+        public int Cols
+        { get; set; }
+
+    }
+
+    public class XmlKeys
+    {
+        [XmlElement(ElementName = "BehaviourKey")]
+        public List<XmlBehaviourKey> BehaviourKeys
+        { get; set; }
+
+        [XmlElement(ElementName = "TextKey")]
+        public List<XmlTextKey> TextKeys
+        { get; set; }
+
+        [XmlElement(ElementName = "LinkKey")]
+        public List<XmlLinkKey> LinkKeys
+        { get; set; }
+
+        [XmlIgnore]
+        public int Count
+        {
+            get
+            {
+                return BehaviourKeys.Count + TextKeys.Count + LinkKeys.Count;
+            }
+        }
+    }
+
+    public class XmlLinkKey : XmlKey
+    {
+
+        public string Link
+        { get; set; }
+
+        [XmlIgnore]
+        public bool ReturnToThisKeyboard
+        { get; set; }
+        
+        [XmlElement("ReturnToThisKeyboard")]
+        public string ReturnToThisKeyboardAsString
+        {
+            get { return this.ReturnToThisKeyboard ? "True" : "False"; }
+            set { this.ReturnToThisKeyboard = XmlUtils.ConvertToBoolean(value); }            
+        }
+    }
+
+    public class XmlBehaviourKey : XmlKey
```

Comment:
Keyboard/Key/Concept renames:
Rename "BehaviourKey" to "ActionKey"
Rename "FunctionKey" to "Action"
Rename "LinkKey" to "ChangeKeyboardKey"
...and "Link" to "DestinationKeyboard"